source/code/providers/Container_HostInventory_Class_Provider.cpp

```diff
@@ -0,0 +1,218 @@
+                {
+                    hostInventoryInstance.NodeRole_value("Master");  
+                }
+                else if(computerName.find("agent") != string::npos)
+                {
+                    hostInventoryInstance.NodeRole_value("Agent");  
+                }
+                else
+                {
+                    hostInventoryInstance.NodeRole_value("Not Orchestrated");  
+                }   
+
+                //Orchestrator type is obtained from the node name. example of node names swarm-manager-vmss_0 for swarm, k8s-master-71E8D996-0 for kubernetes
+                if(computerName.find("swarmm") != string::npos)
+                {
+                    //swarm mode
+                    hostInventoryInstance.OrchestratorType_value("Swarm Mode");  
+                }
+                else if(computerName.find("swarm") != string::npos)
+                {
+                    //swarm
+                    hostInventoryInstance.OrchestratorType_value("Swarm");  
+                }
+                else if(computerName.find("k8") != string::npos)
+                {
+                    //k8
+                    hostInventoryInstance.OrchestratorType_value("Kubernetes");  
+                }
+                else if(computerName.find("dcos") != string::npos)
+                {
+                    //dcos
+                    hostInventoryInstance.OrchestratorType_value("DC/OS");  
+                }
+                else
+                {
+                    hostInventoryInstance.OrchestratorType_value("None");  
+                }   
+                        
+                // Clean up object
+                cJSON_Delete(response[0]);
+            }
+            else
+            {
+                syslog(LOG_WARNING, "API call in Container_HostInventory::InspectHost to get host info returned null");
+            
+            }
+        }
+	//handle all exception, so in any case the omiagent is not brought down
+	catch(exception& e)
+	{
+		syslog(LOG_ERR, "Container_HostInventory %s", e.what());
+	}
+        return hostInventoryInstance;
+    }
+};
+
+Container_HostInventory_Class_Provider::Container_HostInventory_Class_Provider(
+    Module* module) :
+    m_Module(module)
+{
+}
+
+Container_HostInventory_Class_Provider::~Container_HostInventory_Class_Provider()
+{
+}
+
+void Container_HostInventory_Class_Provider::Load(
+        Context& context)
+{
+    context.Post(MI_RESULT_OK);
+}
+
+void Container_HostInventory_Class_Provider::Unload(
+        Context& context)
+{
+    context.Post(MI_RESULT_OK);
+}
+
+void Container_HostInventory_Class_Provider::EnumerateInstances(
+    Context& context,
+    const String& nameSpace,
+    const PropertySet& propertySet,
+    bool keysOnly,
+    const MI_Filter* filter)
+{
+    Container_HostInventory_Class queryResult = ContainerHostInventoryQuery::InspectHost();
+    context.Post(queryResult);
+    context.Post(MI_RESULT_OK);
```

Comment:
Rather than putting the exception handler in InspectHost, the exception handler should go here. In case of an exception, you don't want to post MI_RESULT_OKAY, you want to post MI_RESULT_FAILED.

[SCXCore](https://github.com/Microsoft/SCXcore/blob/master/source/code/providers/SCX_Agent_Class_Provider.cpp#L219) has an example of an exception handler. This is hidden in a macro but, basically, if an exception occurs, then a different result is posted. But if all is good, MI_RESULT_OK is posted.

You do not want to post MI_RESULT_OK if the enumeration was NOT okay ...

See [here](https://github.com/Microsoft/SCXcore/blob/master/source/code/providers/support/scxcimutils.h) for a definition of that macro.